Makefile

```diff
@@ -1,59 +1,61 @@
-TEST_E2E_DIR = ocaml-lsp-server/test/e2e
+.DEFAULT_GOAL := all
 
-build:
-	dune build
-.PHONY: build
+TEST_E2E_DIR = ocaml-lsp-server/test/e2e
 
-fmt:
-	dune build @fmt --auto-promote
-	cd $(TEST_E2E_DIR) && yarn fmt
-.PHONY: fmt
+$(TEST_E2E_DIR)/node_modules:
+	cd $(TEST_E2E_DIR) && yarn install
 
-gen:
-	dune build @cinaps --auto-promote
-	dune build @fmt --auto-promote
-.PHONY: gen
+.PHONY: all
+all:
+	opam exec -- dune build --root . @install
 
+.PHONY: dev
+dev: ## Setup a development environment
+	opam switch create --no-install . ocaml-base-compiler.4.11.1
+	opam install -y dune-release merlin ocamlformat utop ocaml-lsp-server
+	opam install --locked --deps-only --with-test --with-doc -y .
 
-jest-promote:
-	cd $(TEST_E2E_DIR) && yarn promote
-.PHONY: jest-promote
+.PHONY: build
+build: ## Build the project, including non installable libraries and executables
+	opam exec -- dune build --root .
 
-dune-promote:
-	dune promote
-.PHONY: dune-promote
+.PHONY: install
+install: all ## Install the packages on the system
+	opam exec -- dune install --root .
 
-fmt-check:
-	dune build @fmt
-	cd $(TEST_E2E_DIR) && yarn fmt-check
-.PHONY: fmt-check
+.PHONY: lock
+lock: ## Generate the lock files
+	opam lock -y .
 
-lsp-server:
-	dune build @install
-.PHONY: lsp-server
+.PHONY: test
+test: ## Run the unit tests
+	opam exec -- dune build --root . @lsp-fiber/runtest @fiber-unix/runtest @jsonrpc-fiber/runtest
 
-test-e2e: $(TEST_E2E_DIR)/node_modules
-	dune build @install
-	cd $(TEST_E2E_DIR) && dune exec -- yarn test
-.PHONY: test-e2e
+.PHONY: test-template
+test-e2e: $(TEST_E2E_DIR)/node_modules all ## Run the template integration tests
+	cd $(TEST_E2E_DIR) && opam exec -- dune exec -- yarn test
 
-test-ocaml:
-	dune build @lsp-fiber/runtest @fiber-unix/runtest @jsonrpc-fiber/runtest
-.PHONY: test-ocaml
+.PHONY: clean
+clean: ## Clean build artifacts and other generated files
+	opam exec -- dune clean --root .
 
-test: test-ocaml test-e2e
-.PHONY: test
+.PHONY: fmt
+fmt: ## Format the codebase with ocamlformat
+	opam exec -- dune build --root . @fmt --auto-promote
+	cd $(TEST_E2E_DIR) && yarn fmt
 
-clean:
-	dune clean
-.PHONY: clean
+.PHONY: watch
+watch: ## Watch for the filesystem and rebuild on every change
+	opam exec -- dune build ---root . -watch
 
-$(TEST_E2E_DIR)/node_modules:
-	cd $(TEST_E2E_DIR) && yarn install --frozen-lockfile
+.PHONY: utop
+utop: ## Run a REPL and link with the project's libraries
+	opam exec -- dune utop --root . . -- -implicit-bindings
 
-opam-release:
-	dune-release distrib --skip-build --skip-lint --skip-tests --include-submodules
+.PHONY: release
+release: all ## Release on Opam
+	opam exec -- dune-release distrib --skip-build --skip-lint --skip-tests --include-submodules
 	# See https://github.com/ocamllabs/dune-release/issues/206
 	DUNE_RELEASE_DELEGATE=github-dune-release-delegate dune-release publish distrib --verbose
-	dune-release opam pkg
-	dune-release opam submit
```

Comment:
I have `dune-release` in a custom switch so this doesn't work well for me unfortunately. What's the issue with running make like this: `$ opam exec -- make`? It seems to be just as convenient.

Reply:
More typing? 😄  More seriously, I don't have a strong opinion here, I'll just revert this.

Reply:
`$ opam exec -- bash` solves that :)

One thing that we can do to make it work for both of us is to add this line:
```
-include Makefile.dev
```

And then add some code to `Makefile.dev` to turn on the switch. I assume it's possible to set the switch by either customizing the shell command or by reading the environment variables and re-exporting them through make (might be a little difficult)